protocols/floodsub/src/protocol.rs

```diff
@@ -56,10 +62,16 @@ where
     type Error = FloodsubDecodeError;
     type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
 
-    fn upgrade_inbound(self, mut socket: TSocket, _: Self::Info) -> Self::Future {
+    fn upgrade_inbound(self, socket: TSocket, _: Self::Info) -> Self::Future {
         Box::pin(async move {
-            let packet = upgrade::read_length_prefixed(&mut socket, 2048).await?;
-            let rpc = rpc_proto::Rpc::decode(&packet[..]).map_err(DecodeError)?;
+            let mut framed = Framed::<TSocket, prost_codec::Codec<rpc_proto::Rpc>>::new(
+                socket,
+                prost_codec::Codec::new(MAX_MESSAGE_LEN_BYTES),
+            );
+            let rpc = match framed.next().await {
+                Some(v) => v.unwrap_or_default(),
+                _ => panic!(), // Better error here?
+            };
```

Comment:
@thomaseizinger I think I'm really close to the answer on this one, but everything I try looks incorrect (or the compiler complains loudly).

I'm drawing a blank on getting the error in the match. I thought it would be `Err(err) => DecodeError;` but the complaints are loud on that one. It claims it wants to match on `rpc_proto::Error` but including that leads to more errors. :thinking:

Reply:
The key is to use `return`.

Almost everything is an expression Rust so if you don't use control flow here, both branches of the `match` need to evaluate to the same type because it will be stored in the `rpc` variable.

Something like this:

```rust
None => return Err(FloodsubDecodeError)
```

Now, there is also a shortcut for that. `Option` has a combinator `ok_or_else` that allows you to map the `None` path to an `Err` which allows you to use `?`:

```rust
let result = stream.next().await.unwrap_or_else(|| FloodsubDecodeError)?;
```

(Examples subject to minor tweaking to properly construct the error but you get the idea.)